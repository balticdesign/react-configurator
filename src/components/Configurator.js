import React from 'react';
import FacetSection from './FacetSection';
import Price from './Price';

const Configurator = props => (
  <div>
    <div className="container">
      <div className="row">
        <div className="col s12">
          <ul className="collapsible" data-collapsible="accordion">
            {
              props.facets.map(facet => (
                <li key={facet.id}>
                  <FacetSection
                    facet={facet}
                    isOpen={(facet.id === props.ui.selectedFacetId)}
                    selection={props.getOptionDescription(facet.id)}
                    onClick={() => props.onSectionChange(facet.id)}
                    onOptionChosen={props.onOptionChosen}
                  />
                </li>
              ))
            }
          </ul>
        </div>
      </div>
      <div className="row">
        <div className="col offset-s7 s5">
          <Price price={props.getConfiguredPrice()} />
        </div>
      </div>
    </div>
  </div>
);

Configurator.propTypes = {
  facets: React.PropTypes.arrayOf(React.PropTypes.object),
  onOptionChosen: React.PropTypes.func,
  getConfiguredPrice: React.PropTypes.func,
};

export default Configurator;
