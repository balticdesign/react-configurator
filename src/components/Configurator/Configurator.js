import React from 'react';
import PropTypes from 'prop-types';
import FacetSection from '../FacetSection';
import Price from '../Price';

const Configurator = props => (
  <div>
    <div className="container">
      <div className="row">
        <div className="col s12">
          <ul className="collapsible" data-collapsible="accordion">
            {
              Object.keys(props.facets).map(facetKey => (
                <li key={facetKey}>
                  <FacetSection
                    facet={props.facets[facetKey]}
                    facetKey={facetKey}
                    selectedOption={props.selections[facetKey]}
                    isOpen={(facetKey === props.ui.selectedFacetId)}
                    selection={props.getOptionDescription(facetKey)}
                    onClick={() => props.onSectionChange(facetKey)}
                    onOptionChosen={props.onOptionChosen}
                  />
                </li>
              ))
            }
          </ul>
        </div>
      </div>
      <div className="row">
        <div className="col offset-s7 s5">
          <Price price={props.getConfiguredPrice()} />
        </div>
      </div>
    </div>
  </div>
);

// Looks like a bug with eslint or plugin for unused props and stateless components
/* eslint-disable react/no-unused-prop-types */
Configurator.propTypes = {
  facets: PropTypes.object,
  selections: PropTypes.object,
  onOptionChosen: PropTypes.func,
  getConfiguredPrice: PropTypes.func,
  getOptionDescription: PropTypes.func,
  onSectionChange: PropTypes.func,
  ui: PropTypes.object,
};
/* eslint-enable */

export default Configurator;
