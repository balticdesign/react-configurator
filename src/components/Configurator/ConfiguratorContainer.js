import { connect } from 'react-redux';
import { changeSelectedFacet, optionSelected } from '../../actions/index';
import { getOptionDescription, getConfiguredPrice } from '../../reducers/facet-selectors';
import Configurator from './Configurator';

const mapStateToProps = state => (
  {
    facets: state.facets.data,
    selections: state.facets.selections,
    ui: state.ui,
    getOptionDescription: facetId => (getOptionDescription(state.facets, facetId)),
    getConfiguredPrice: () => (getConfiguredPrice(state.facets)),
  }
);

const mapDispatchToProps = dispatch => (
  {
    onSectionChange: facetId => dispatch(changeSelectedFacet(facetId)),
    onOptionChosen: (facetId, optionId) => dispatch(optionSelected(facetId, optionId)),
  }
);

const ConfiguratorContainer = connect(
  mapStateToProps,
  mapDispatchToProps)(Configurator);

export default ConfiguratorContainer;
