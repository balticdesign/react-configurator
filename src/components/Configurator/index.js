import ConfiguratorContainer from './ConfiguratorContainer';

export default ConfiguratorContainer;
