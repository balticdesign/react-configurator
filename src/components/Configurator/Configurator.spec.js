import React from 'react';
import renderer from 'react-test-renderer';
import Enzyme, { mount } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import Configurator from './Configurator';
import testFacets from '../../Utils/testUtils';

Enzyme.configure({ adapter: new Adapter() });

const uiState = {
  selectedFacetId: 'size',
};

describe('Configurator tests', () => {
  it('basic snapshot', () => {
    const tree = renderer.create(
      <Configurator
        facets={testFacets.data}
        selections={testFacets.selections}
        ui={uiState}
        getOptionDescription={facetId => `*** ${facetId}`}
        getConfiguredPrice={() => 500}
      />,
    );
    expect(tree).toMatchSnapshot();
  });

  it('should render a section for each facet', () => {
    const wrapper = mount(
      <Configurator
        facets={testFacets.data}
        selections={testFacets.selections}
        ui={uiState}
        getOptionDescription={facetId => `*** ${facetId}`}
        getConfiguredPrice={() => 500}
      />,
    );

    expect(wrapper.find('.collapsible > li')).toHaveLength(Object.keys(testFacets.data).length);
  });

  it('should render with selected facet open', () => {
    const wrapper = mount(
      <Configurator
        facets={testFacets.data}
        selections={testFacets.selections}
        ui={uiState}
        getOptionDescription={facetId => `*** ${facetId}`}
        getConfiguredPrice={() => 500}
      />,
    );

    expect(wrapper.find('.collapsible > li').at(0).props().children.props.isOpen).toBe(true);
    expect(wrapper.find('.collapsible > li').at(1).props().children.props.isOpen).toBe(false);
    expect(wrapper.find('.collapsible > li').at(2).props().children.props.isOpen).toBe(false);
  });

  it('should call back on change of section', () => {
    const sectionToClick = 2;
    const onSectionClick = jest.fn();
    const wrapper = mount(
      <Configurator
        facets={testFacets.data}
        selections={testFacets.selections}
        ui={uiState}
        getOptionDescription={facetId => `*** ${facetId}`}
        getConfiguredPrice={() => 500}
        onSectionChange={onSectionClick}
      />,
    );

    const section = wrapper.find('.collapsible-header').at(sectionToClick - 1);
    section.simulate('click');

    expect(onSectionClick.mock.calls).toHaveLength(1);
    expect(onSectionClick.mock.calls[0][0]).toBe('groupset');
  });
});
