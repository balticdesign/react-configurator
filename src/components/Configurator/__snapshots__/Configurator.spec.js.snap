// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Configurator tests basic snapshot 1`] = `
<div>
  <div
    className="container"
  >
    <div
      className="row"
    >
      <div
        className="col s12"
      >
        <ul
          className="collapsible"
          data-collapsible="accordion"
        >
          <li>
            <div>
              <div
                className="collapsible-header"
                onClick={[Function]}
                role="presentation"
              >
                <i
                  className="material-icons"
                />
                <span
                  className="facet-section--notSelected"
                >
                  *** size
                </span>
              </div>
              <div
                className="collapsible-body"
                style={
                  Object {
                    "display": "block",
                  }
                }
              >
                <p />
                <ul
                  className="facet-options"
                >
                  <li
                    className="facet-options__body"
                  >
                    <input
                      checked={false}
                      id="section-size-0"
                      name="section-size"
                      onChange={[Function]}
                      type="radio"
                    />
                    <label
                      htmlFor="section-size-0"
                    >
                      Facet1Desc1
                      
                    </label>
                  </li>
                  <li
                    className="facet-options__body"
                  >
                    <input
                      checked={false}
                      id="section-size-1"
                      name="section-size"
                      onChange={[Function]}
                      type="radio"
                    />
                    <label
                      htmlFor="section-size-1"
                    >
                      Facet1Desc2
                      
                    </label>
                  </li>
                  <li
                    className="facet-options__body"
                  >
                    <input
                      checked={false}
                      id="section-size-2"
                      name="section-size"
                      onChange={[Function]}
                      type="radio"
                    />
                    <label
                      htmlFor="section-size-2"
                    >
                      Facet1Desc3
                       (+£100)
                    </label>
                  </li>
                </ul>
              </div>
            </div>
          </li>
          <li>
            <div>
              <div
                className="collapsible-header"
                onClick={[Function]}
                role="presentation"
              >
                <i
                  className="material-icons"
                />
                <span
                  className="facet-section--notSelected"
                >
                  *** groupset
                </span>
              </div>
              <div
                className="collapsible-body"
                style={
                  Object {
                    "display": "none",
                  }
                }
              >
                <p />
                <ul
                  className="facet-options"
                >
                  <li
                    className="facet-options__body"
                  >
                    <input
                      checked={false}
                      id="section-groupset-0"
                      name="section-groupset"
                      onChange={[Function]}
                      type="radio"
                    />
                    <label
                      htmlFor="section-groupset-0"
                    >
                      Facet2Desc1
                      
                    </label>
                  </li>
                  <li
                    className="facet-options__body"
                  >
                    <input
                      checked={false}
                      id="section-groupset-1"
                      name="section-groupset"
                      onChange={[Function]}
                      type="radio"
                    />
                    <label
                      htmlFor="section-groupset-1"
                    >
                      Facet2Desc2
                       (+£200)
                    </label>
                  </li>
                  <li
                    className="facet-options__body"
                  >
                    <input
                      checked={false}
                      id="section-groupset-2"
                      name="section-groupset"
                      onChange={[Function]}
                      type="radio"
                    />
                    <label
                      htmlFor="section-groupset-2"
                    >
                      Facet2Desc3
                      
                    </label>
                  </li>
                </ul>
              </div>
            </div>
          </li>
          <li>
            <div>
              <div
                className="collapsible-header"
                onClick={[Function]}
                role="presentation"
              >
                <i
                  className="material-icons"
                />
                <span
                  className="facet-section--notSelected"
                >
                  *** colour
                </span>
              </div>
              <div
                className="collapsible-body"
                style={
                  Object {
                    "display": "none",
                  }
                }
              >
                <p />
                <ul
                  className="facet-options"
                >
                  <li
                    className="facet-options__body"
                  >
                    <input
                      checked={false}
                      id="section-colour-0"
                      name="section-colour"
                      onChange={[Function]}
                      type="radio"
                    />
                    <label
                      htmlFor="section-colour-0"
                    >
                      Facet3Desc1
                       (+£300)
                    </label>
                  </li>
                  <li
                    className="facet-options__body"
                  >
                    <input
                      checked={false}
                      id="section-colour-1"
                      name="section-colour"
                      onChange={[Function]}
                      type="radio"
                    />
                    <label
                      htmlFor="section-colour-1"
                    >
                      Facet3Desc2
                      
                    </label>
                  </li>
                  <li
                    className="facet-options__body"
                  >
                    <input
                      checked={false}
                      id="section-colour-2"
                      name="section-colour"
                      onChange={[Function]}
                      type="radio"
                    />
                    <label
                      htmlFor="section-colour-2"
                    >
                      Facet3Desc3
                      
                    </label>
                  </li>
                </ul>
              </div>
            </div>
          </li>
        </ul>
      </div>
    </div>
    <div
      className="row"
    >
      <div
        className="col offset-s7 s5"
      >
        <div
          className="card purple darken-4"
        >
          <div
            className="card-content white-text"
          >
            <span
              className="card-title right-align"
            >
              Price £
              500
            </span>
            <p>
              Interest-free credit is available on this item
            </p>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
